File: output_at_5/qwen3/sample_3/c/c/Struct/deposit.c
#include "stdlib.h"

/*@
predicate account(struct account *acc, int balance) = acc->balance |-> balance &*& acc != 0;
@*/

struct account
{
    int balance;
};

//@ requires true;
//@ ensures \result != 0 ==> account(\result, 0);
struct account *create_account()
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    return myAccount;
}

//@ requires account(myAccount, ?bal);
//@ requires amount >= 0;
//@ ensures account(myAccount, bal + amount);
void account_deposit(struct account *myAccount, int amount)
{
    myAccount->balance += amount;
}

//@ requires account(myAccount, ?bal);
//@ ensures \result == bal;
int account_get_balance(struct account *myAccount)
{
    int result = myAccount->balance;
    return result;
}

//@ requires account(myAccount, ?bal);
//@ requires newBalance >= 0;
//@ ensures account(myAccount, newBalance);
void account_set_balance(struct account *myAccount, int newBalance)
{
    myAccount->balance = newBalance;
}

//@ requires account(myAccount, ?bal);
//@ ensures true;
void account_dispose(struct account *myAccount)
{
    free(myAccount);
}

int main()
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output_20260321_064953/claude-opus/rust/Struct/deposit.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Account {
    balance: i32,
}

impl Account {

    //@ pred account_ptr(?p;) = p::Account_bal(p, ?b);
    
    //@ predicate Account_bal(*mut Account acc, int bal) = acc |-> Account { balance: bal };
    
    //@ requires true;
    //@ ensures result |-> Account { balance: 0 };
    unsafe fn create() -> *mut Account
    
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).balance = 0;
        return my_account;
    }

    //@ requires my_account |-> Account { balance: _ };
    //@ ensures my_account |-> Account { balance: _ };
    unsafe fn get_balance(my_account: *mut Account) -> i32
    
    {
        return (*my_account).balance;
    }

    //@ requires my_account |-> Account { balance: _ };
    //@ ensures my_account |-> Account { balance: new_balance };
    unsafe fn set_balance(my_account: *mut Account, new_balance: i32)
    
    {
        (*my_account).balance = new_balance;
    }

    //@ requires my_account |-> Account { balance: ?b };
    //@ ensures my_account |-> Account { balance: b + amount };
    unsafe fn deposit(my_account: *mut Account, amount: i32)
    
    {
        (*my_account).balance += amount;
    }

    //@ requires my_account |-> Account { balance: _ };
    //@ ensures true;
    unsafe fn dispose(my_account: *mut Account)
    
    {
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }

}

fn main()

{
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}